arduino/smartcar/smartcar.ino

```diff
@@ -1,9 +1,44 @@
 #include <Smartcar.h>
 
+//for the ultrasonic sensor
+const unsigned int MAX_DISTANCE = 80;
+//sends impulse
+const int TRIGGER_PIN = 6;
+//receives reply
+const int ECHO_PIN = 7;
+
+ArduinoRuntime arduinoRuntime;
+BrushedMotor leftMotor(arduinoRuntime,smartcarlib::pins::v2::leftMotorPins);
+BrushedMotor rightMotor(arduinoRuntime,smartcarlib::pins::v2::rightMotorPins);
+DifferentialControl control(rightMotor, leftMotor);
+SR04 frontUltraSonic(arduinoRuntime, TRIGGER_PIN,
+ECHO_PIN, MAX_DISTANCE);
+
 void setup() {
-  // put your setup code here, to run once:
+  Serial.begin(9600);
 }
 
 void loop() {
-  // put your main code here, to run repeatedly:
+  Serial.println(frontUltraSonic.getDistance());
+  obstacleAvoidance();
 }
+
+void obstacleAvoidance(){
+  //since if the sensors pick up nothing the distance is 0
+  /**
+   * @brief if the distance is greater than 0 and less than 35
+   * meaning it has picked up an obstacle but waits until it has reached
+   * a point before crashing, then it stops.
+   *
+   * Value for min distance to obstalce = 35;
+   * 
+   * @else drive normally
+   * 
+   */
+  if(frontUltraSonic.getDistance() > 0 && frontUltraSonic.getDistance() < 70){
```

Comment:
What happens if the first call to `getDistance` returns `71` and the second one returns `0`?

Reply:
It will keep driving as the condition is not satisfied

Reply:
It will be TRUE and FALSE which is FALSE

Reply:
71 means that it continues driving
0 means it also continues driving since only one condition is stasfied (sensor did not pick anything up)

Reply:
Read the problem statement again and look carefully at L38:

> What happens if the first call to getDistance returns 71 and the second one returns 0?

The car stops while it should not.

Reply:
so as soon as the sensor picks up an obstacle (distance > 0) and its distance is less than the maximum distance to obstacle (distance = 70) it will stop

Reply:
I'll let you think of it more carefully :smirk:

Reply:
So you are saying more like if the first mention of getDistance() reads 71 and the second getDistance() reads 0 --> in that case the car would stop yes since both statements would be true (71 > 0 AND 0<70 == TRUE)

Reply:
Yes, this is what I said indeed. Is it the correct behavior? No obstacle closer than 70 centimetres was detected, and yet the car stopped.

Reply:
This is very likely to happen in real life. In the emulator maybe, but less likely as it's more "perfect" than the real world.
Measurements are noisy, especially at longer distances or objects at "inconvenient" angles. There's no guarantee that two measurements will be the same, no matter how close they occur.

Reply:
Has a more robust check whether the distance is actually there and not 'noise'

Reply:
Nope, it has exactly the same problem.

Reply:
Yeah i see just reverse the pairing you gave me and it will stop 😆

Reply:
Am I right in thinking that successive calls to getDistance() is not the answer you are looking for due to this noise?

Reply:
Could work
- There is only one call meaning noise not being as prevalent
- The max distance would be whatever the first measurement is for the sensor
- Drives until an object is even remotely picked up, then stops
- in the case of getting 71 and 0 and then 71 again it would stop, drive, then stop

if(frontUltraSonic.getDistance() > 0){
      control.setSpeed(0);
  }
  else{
    control.setSpeed(40);
  }

Reply:
Wouldn't be the ideal or most specific solution but it stops at any given object regardless of distance

Reply:
The solution you propose is not just "not ideal" but wrong too as it doesn't satisfy the requirements of the assignment. :)
Keep thinking or actually stop overthinking it. :exploding_head:

Reply:
Not using the loop() method I could propose this as simple as possible
- This satisfies the requirement of a specified stopping distance
- Given that requirement any distance less or more should not be covered by the system
- This is me thinking outside the loop() box and outside my previous code with >,< comparisons
- I understand that if the distance != 60 then it will hit the obstacle but that is not covered in the requirement for the assignment (meaning values of 59 and lower will lead to a crash) however it states:  the Smartcar move forward until it detects an obstacle at a distance of your choosing -- which it does
  
while(true){
    //if the sensor returns the distance of our choosing then stop
    if(frontUltraSonic.getDistance() == 60){
      control.setSpeed(0);
      break;
    }
    //if distance is not the one chose then we stop
    else{
      control.setSpeed(20);
    }
  }

Reply:
No, but feel free to ask @DIT113-V22/group-01 for input.

Reply:
I have asked but no response on their part sadly. I will try again

Reply:
Looks like I was running in the wrong direction the whole time... 🤔 

auto distance = frontUltraSonic.getDistance();
  if(distance > 0 && distance < 60){
    control.setSpeed(0);
    delay(100);
  }
  else{
    control.setSpeed(20);
  }

Reply:
We discussed it on whatsapp and were able to solve it 🙌